Clarify docs for Flow.ask generic parameter

raboof opened the pull request.

raboof committed:
Clarify docs for Flow.ask generic parameter

johanandren commented on akka-docs/src/main/paradox/stream/operators/Source-or-Flow/ask.md:
```diff
@@ -15,7 +15,7 @@ Use the `ask` pattern to send a request-reply message to the target `ref` actor.
 If any of the asks times out it will fail the stream with a @apidoc[AskTimeoutException].
 
-The `mapTo` class parameter is used to cast the incoming responses to the expected response type.
+The `S` generic parameter is used to cast the responses from the actor to the expected outgoing flow type.
```
"If a different type of answer is sent this will lead to the future being failed with a class cast exception" or something like that?

raboof replied:
Good one - I wonder if we should throw a custom `ClassCastException` making it more clear where it's coming from

johanandren replied:
I think that is a good idea 👍

patriknw replied:
the Scaladoc also mention "Do not forget to include the expected response type...", which can be good to have here as well (for `@scala`)

patriknw replied:
the text should be different for `@scala` and `@java`. for java there is `mapTo` parameter

raboof committed:
Java/scala switch

raboof committed:
Throw specific ClassCastException in Flow.ask

Benchmark seems too noisy to draw any real conclusions, but seems not terrible:

```
pre:
Score        Error  Units
[info] AskBenchmark.mapAsync              1    false  thrpt    3  323497.536 ±  38733.024  ops/s
[info] AskBenchmark.mapAsync              1     true  thrpt    3  327001.647 ± 382202.906  ops/s
[info] AskBenchmark.mapAsync              4    false  thrpt    3  457725.806 ±  67583.859  ops/s
[info] AskBenchmark.mapAsync              4     true  thrpt    3  451728.770 ±  74886.315  ops/s

post:
[info] Benchmark              (parallelism)  (spawn)   Mode  Cnt       Score        Error  Units
[info] AskBenchmark.mapAsync              1    false  thrpt    3  338509.183 ± 160893.556  ops/s
[info] AskBenchmark.mapAsync              1     true  thrpt    3  345710.807 ± 505445.147  ops/s
[info] AskBenchmark.mapAsync              4    false  thrpt    3  737276.613 ±   3306.911  ops/s
[info] AskBenchmark.mapAsync              4     true  thrpt    3  697960.417 ± 454528.184  ops/s

[info] Benchmark              (parallelism)  (spawn)   Mode  Cnt       Score        Error  Units
[info] AskBenchmark.mapAsync              1    false  thrpt    3  349120.202 ± 111625.074  ops/s
[info] AskBenchmark.mapAsync              1     true  thrpt    3  440218.776 ± 817228.408  ops/s
[info] AskBenchmark.mapAsync              4    false  thrpt    3  665269.418 ±  10502.449  ops/s
[info] AskBenchmark.mapAsync              4     true  thrpt    3  471313.943 ±  35173.757  ops/s

```

raboof commented:
benchmark seems to noisy to draw any real conclusions:

pre:
```
Score        Error  Units
[info] AskBenchmark.mapAsync              1    false  thrpt    3  323497.536 ±  38733.024  ops/s
[info] AskBenchmark.mapAsync              1     true  thrpt    3  327001.647 ± 382202.906  ops/s
[info] AskBenchmark.mapAsync              4    false  thrpt    3  457725.806 ±  67583.859  ops/s
[info] AskBenchmark.mapAsync              4     true  thrpt    3  451728.770 ±  74886.315  ops/s
```
post:
```
[info] Benchmark              (parallelism)  (spawn)   Mode  Cnt       Score        Error  Units
[info] AskBenchmark.mapAsync              1    false  thrpt    3  338509.183 ± 160893.556  ops/s
[info] AskBenchmark.mapAsync              1     true  thrpt    3  345710.807 ± 505445.147  ops/s
[info] AskBenchmark.mapAsync              4    false  thrpt    3  737276.613 ±   3306.911  ops/s
[info] AskBenchmark.mapAsync              4     true  thrpt    3  697960.417 ± 454528.184  ops/s

[info] Benchmark              (parallelism)  (spawn)   Mode  Cnt       Score        Error  Units
[info] AskBenchmark.mapAsync              1    false  thrpt    3  349120.202 ± 111625.074  ops/s
[info] AskBenchmark.mapAsync              1     true  thrpt    3  440218.776 ± 817228.408  ops/s
[info] AskBenchmark.mapAsync              4    false  thrpt    3  665269.418 ±  10502.449  ops/s
[info] AskBenchmark.mapAsync              4     true  thrpt    3  471313.943 ±  35173.757  ops/s
```

johanandren commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1017,7 +1017,11 @@ trait FlowOps[+Out, +Mat] {
+      }
+      .map {
+        case e: S => e
+        case o => throw new ClassCastException(s"'Flow.ask' failed: expected response of type [$tag], got [${o.getClass}]")
```
Not `${tag.runtimeClass}` ?

johanandren commented on akka-bench-jmh/src/main/scala/akka/stream/AskBenchmark.scala:
```diff
@@ -0,0 +1,92 @@
+    }
+  }
+
+}
```
Yea but how fast is it? ;)

raboof committed:
Formatting, use tag.runtimeClass

johanandren reviewed (approved):
LGTM

patriknw reviewed (approved):
LGTM

patriknw commented:
@raboof is this important enough to backport?

raboof commented:
I think it's fine just for 2.6
